core/src/main/java/com/transferwise/common/gracefulshutdown/GracefulShutdowner.java

```diff
@@ -82,7 +82,11 @@ public void stop() {
 
         long start = System.currentTimeMillis();
         List<GracefulShutdownStrategy> redLightStrategies = new ArrayList<>(strategies);
-        while (System.currentTimeMillis() - start < properties.getShutdownTimeoutMs()) {
+
+        // slightly increase timeout here
+        // now strategies could safely try to shut down within GracefulShutdownProperties.getShutdownTimeoutMs()
+        int safeShutdownTimeoutMs = properties.getShutdownTimeoutMs() + 500;
```

Comment:
Maybe ideally we need to additionally multiply this by 2 if strategies is not yet ready to shutdown. It could work something like this:

1. Wait for ShutdownTimeoutMs for graceful shutdown.
2. If strategies is ready - exit.
3. If strategies is not ready run force shutdown in strategies and wait for additional ShutdownTimeoutMs.

Currently shutdown logic in `ExecutorServiceGracefulShutdownStrategy` just split ShutdownTimeoutMs by parts:

1. Graceful shut down timeout is 5/6 of ShutdownTimeoutMs
2. Force shutdown timeout is 1/6 of ShutdownTimeoutMs

Reply:
As it seems you are proposing making the shutdown multi phase? What do you mean by force shutdown in this case? You mean extending the `GracefulShutdownStrategy` interface with another lifecycle method? 
I understand that for ExecutorServiceGracefulShutdownStrategy it's encapsulated into the current logic itself.

Reply:
About this point:
`Allow to send new tasks for 5/10 of ShutdownTimeoutMs`
What is this based on? This only makes sense if before/during that delay endpoints are actually closed as otherwise it's just pushing the problem down the line. For example, you gracefully service requests with come in right after the shutdown was initiated, but if a request comes in right before the delay is up then it will still get an exception.
Currently there is also this logic [here](https://github.com/transferwise/tw-graceful-shutdown/blob/implement-executor-service-shutdown-strategy/core/src/main/java/com/transferwise/common/gracefulshutdown/GracefulShutdowner.java#L81) which waits for client reaction time. More specifically it's aligned with envoy timeout which stops sending requests to that pod which is shutting down. I wonder can we utilise that logic, as the idea is the same.

Reply:
You right, it will be safer to use `ClientsReactionTimeMs` + maybe 500 ms instead of `5/10 of ShutdownTimeoutMs`. . Right now `5/10 of ShutdownTimeoutMs` = `ClientsReactionTimeMs`, but it may not be like this in the future. I will change logic according to your comment.

Reply:
> As it seems you are proposing making the shutdown multi phase? What do you mean by force shutdown in this case? You mean extending the `GracefulShutdownStrategy` interface with another lifecycle method? I understand that for `ExecutorServiceGracefulShutdownStrategy` it's encapsulated into the current logic itself.

Encapsulating might be sufficient. I personally don't see benefits of introducing new stage as it will introduce more complexity. The main point - not close app right away if strategies is not ready and allow to wait for at least additional `ShutdownTimeoutMs`.